### PR TITLE
Adds keyboard shell

### PR DESCRIPTION
**Pull request by robotduinomm**
## About The Pull Request

 Adds keyboard shell. 
- Works very similar to the thought listener component for BCIs
- It opens up an input panel when you use it (tgui_input_text)
- Players with the illiterate quirk cannot use this shell. When they try to use it they get a warning message.
- It's printed out as assembled (like compact remote or controller)
- It can fit in a bag but not in a box
- Has small capacity (SHELL_CAPACITY_SMALL)
- Available after researching the Advanced Shells tech node
- Can be printed from the component printer or the protolathe
- Costs 2000 glass and 10000 iron sheets to print (slightly more than the controller shell)
- Has 3 outputs. First one returns the entity who used the shell, second is the text that entity wrote and submitted, third one is triggered after the input window is closed. (by submitting, cancelling or just hitting the X button)

Icons (screenhots from the dream maker)

 Assembled        |  Unassembled   
:-------------------------:|:-------------------------:
aatatatapng |  at

Simple use (In game screenshot):
aaScreenshot (529)

## Why It's Good For The Game

Currently, there are a few options when a user wants to input strings to a shell:
- You can use the voice activator component but it's really hard to use cause it listens all the time and can pick up any speech.
- Or you can use the thought listener component which requires a BCI shell but this is not practical most of the time since only the person that has the shell inside of them can use it.

Circuitry lacks a text input method that is reliable and practical.

With the keyboard shell you can open up a text input window and use it without anyone hearing or messing up with your input.
It's also easy to carry and diffrent players can use a single shell unlike BCI
Some use cases might be:
- Making a device that allows players to speak without using their own voice (you just connect the shells output to a speech component and ta-da)
- You can use it to adjust the input values of the components that has a broad range of possible inputs (eg. sound level for the sound emitter component)
- Or you can just simply use it as an advanced version of compact remote or controller (eg. If the text entered is this do this) 

## Changelog

:cl:
add: Add keyboard shell. A shell that allows the user to input a string. Available after researching the Advanced Shells tech node.
imageadd: added 2 icons for the keyboard shell. One unassembled and one assembled.
/:cl:

### CODE REVIEW

**Comment by YusufEmirKoroglu** [comment]
This is my first time making a pull request so any feedback would be appreciated. Thanks.

**Comment by Elli-Skala** [comment]
Oh maybe make it have a check for if the person typing is illiterate

**Comment by YusufEmirKoroglu** [comment]
What causes Run Linters to fail. How can I fix it.

**Comment by SuperSlayer0** [comment]
> What causes Run Linters to fail. How can I fix it.

code/modules/wiremod/shell/keyboard.dm, line 39, column 61:
error: /obj/item/circuit_component/keyboard_shell/proc/send_trigger sets SpacemanDMM_should_not_sleep but calls blocking proc /proc/tgui_input_text
- 40:2: SpacemanDMM_should_not_sleep set here
- 48:17: /proc/tgui_input_text() called here
- code/modules/tgui_input/text.dm:36:12: input() called here
- code/modules/tgui_input/text.dm:38:12: input() called here

**Comment by Watermelon914** [comment]
very nice 👍

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/modules/research/techweb/all_nodes.dm [hunk 1]
```diff
@@ -738,6 +738,7 @@
 		"dispenser_shell",
 		"door_shell",
 		"gun_shell",
+		"keyboard_shell",
```
Comment (Watermelon914) [inline code/modules/research/techweb/all_nodes.dm:741]: needs to be placed in alphabetical order in line with everything else

Comment (YusufEmirKoroglu) [reply code/modules/research/techweb/all_nodes.dm:741]: I think it's already in alphabetical order. Am I missing something ?

Comment (Watermelon914) [reply code/modules/research/techweb/all_nodes.dm:741]: me dumb

**Comment by Watermelon914** [comment]
it'd be cool if you could make both a handmade version and a physical version you can place on a table and click on to interact with. Just a suggestion, but I'll merge this in 30 minutes if you decide not to

File: code/modules/wiremod/shell/keyboard.dm [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+/obj/item/circuit_component/keyboard_shell/register_shell(atom/movable/shell)
+	RegisterSignal(shell, COMSIG_ITEM_ATTACK_SELF, .proc/send_trigger)
+
+/obj/item/circuit_component/keyboard_shell/unregister_shell(atom/movable/shell)
+	UnregisterSignal(shell, COMSIG_ITEM_ATTACK_SELF)
+
+/obj/item/circuit_component/keyboard_shell/proc/send_trigger(atom/source, mob/user)
+	SIGNAL_HANDLER
+	check_illiterate(user)
```
Comment (Watermelon914) [inline code/modules/wiremod/shell/keyboard.dm:41]: just move this check into the use_keyboard proc, you don't need another proc for this

File: code/modules/wiremod/shell/keyboard.dm [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+/obj/item/circuit_component/keyboard_shell/proc/use_keyboard(mob/user)
+	if(!illiterate)
+		var/message = tgui_input_text(user, "Input your text", "Keyboard")
+		entity.set_output(user)
+		output.set_output(message)
+		signal.set_output(COMPONENT_SIGNAL)
+
+/obj/item/circuit_component/keyboard_shell/proc/check_illiterate(mob/user)
+	if(HAS_TRAIT(user, TRAIT_ILLITERATE))
+		illiterate = TRUE
```
Comment (Watermelon914) [inline code/modules/wiremod/shell/keyboard.dm:53]: this variable isn't needed, just move this check into the use_keyboard proc

File: code/modules/wiremod/shell/keyboard.dm [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		new /obj/item/circuit_component/keyboard_shell()
+	), SHELL_CAPACITY_SMALL)
+
+/obj/item/circuit_component/keyboard_shell
+	display_name = "Keyboard Shell"
+	desc = "A handheld shell that allows the user to input a string. Use the shell in hand to open the input panel."
+
+	var/datum/port/output/signal
+	var/datum/port/output/entity
+	var/datum/port/output/output
```
Comment (Watermelon914) [inline code/modules/wiremod/shell/keyboard.dm:24]: Document these ports, explain what they do in brief detail

**Comment by YusufEmirKoroglu** [comment]
> it'd be cool if you could make both a handmade version and a physical version you can place on a table and click on to interact with. Just a suggestion, but I'll merge this in 30 minutes if you decide not to

That's a great idea. I can work on it but I don't know when I would be able to finish so can you still merge after I made the last 3 changes you requested.

File: code/modules/wiremod/shell/keyboard.dm [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	/// The string, entity typed and submitted
+	var/datum/port/output/output
+
+/obj/item/circuit_component/keyboard_shell/populate_ports()
+	entity = add_output_port("User", PORT_TYPE_ATOM)
+	output = add_output_port("Message", PORT_TYPE_STRING)
+	signal = add_output_port("Signal", PORT_TYPE_SIGNAL)
+
+/obj/item/circuit_component/keyboard_shell/register_shell(atom/movable/shell)
+	RegisterSignal(shell, COMSIG_ITEM_ATTACK_SELF, .proc/send_trigger)
```
Comment (Watermelon914) [inline code/modules/wiremod/shell/keyboard.dm:35]: call parent here (although parent has no code, it's standard for all components to do this)

File: code/modules/wiremod/shell/keyboard.dm [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+/obj/item/circuit_component/keyboard_shell/unregister_shell(atom/movable/shell)
+	UnregisterSignal(shell, COMSIG_ITEM_ATTACK_SELF)
```
Comment (Watermelon914) [inline code/modules/wiremod/shell/keyboard.dm:38]: call parent at the end of your code here (for the same reasons as above)

File: code/modules/wiremod/shell/keyboard.dm (hunk 5 above)
Comment (Watermelon914) [reply code/modules/wiremod/shell/keyboard.dm:35]:
```suggestion
/obj/item/circuit_component/keyboard_shell/register_shell(atom/movable/shell)
	. = ..()
	RegisterSignal(shell, COMSIG_ITEM_ATTACK_SELF, .proc/send_trigger)
```

File: code/modules/wiremod/shell/keyboard.dm (hunk 6 above)
Comment (Watermelon914) [reply code/modules/wiremod/shell/keyboard.dm:38]:
```suggestion
/obj/item/circuit_component/keyboard_shell/unregister_shell(atom/movable/shell)
	UnregisterSignal(shell, COMSIG_ITEM_ATTACK_SELF)
	return ..()
```

**Review by Watermelon914 — CHANGES_REQUESTED** [review]
last nitpicks and then it should be good for merge

**Comment by YusufEmirKoroglu** [comment]
Thank you.

**Comment by YusufEmirKoroglu** [comment]
> last nitpicks and then it should be good for merge

Hey are there any changes required to merge.

**Pull request merged**